global: Fix zero-length array GNU C extension usage

stephanosio opened the pull request:
Declaring zero-length arrays is allowed in GNU C as an extension and is
not part of the C language standards.

The ISO C99 has an equivalent mechanism called "flexible array member"
that serves the same purpose with slightly different syntax and
semantics:

* Flexible array members are written as `contents[]` without the `0`.
* Flexible array members have incomplete type and `sizeof` operator may
  not be applied.
* Flexible array members may only appear as the last member of a
  `struct` that is otherwise non-empty.
* A structure containing a flexible array member, or a union containing
  such a structure (possibly recursively), may not be a member of a
  structure or an element of an array.

The ISO C99 flexible array member mechanism is preferable to the GNU
zero-length array extension because:

1. The former is a standard mechanism and therefore more portable.
2. The former syntatically enforces that flexible array members must be
   the last member of a struct, whereas the latter declares any access
   to interior zero-length array members as undefined, which is highly
   error-prone.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
global: Fix zero-length array GNU C extension usage

Declaring zero-length arrays is allowed in GNU C as an extension and is
not part of the C language standards.

The ISO C99 has an equivalent mechanism called "flexible array member"
that serves the same purpose with slightly different syntax and
semantics:

* Flexible array members are written as `contents[]` without the `0`.
* Flexible array members have incomplete type and `sizeof` operator may
  not be applied.
* Flexible array members may only appear as the last member of a
  `struct` that is otherwise non-empty.
* A structure containing a flexible array member, or a union containing
  such a structure (possibly recursively), may not be a member of a
  structure or an element of an array.

The ISO C99 flexible array member mechanism is preferable to the GNU
zero-length array extension because:

1. The former is a standard mechanism and therefore more portable.
2. The former syntatically enforces that flexible array members must be
   the last member of a struct, whereas the latter declares any access
   to interior zero-length array members as undefined, which is highly
   error-prone.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

stephanosio commented:
From https://gcc.gnu.org/onlinedocs/gcc/Zero-Length.html:

> Declaring zero-length arrays is allowed in GNU C as an extension.

> The preferred mechanism to declare variable-length types like struct line above is the ISO C99 flexible array member

> Declaring zero-length arrays in other contexts, including as interior members of structure objects or as non-member objects, is discouraged. Accessing elements of zero-length arrays declared in such contexts is undefined and may be diagnosed.

joerchan reviewed (approved):
Thanks. LGTM

pabigot reviewed (approved):
Looks like this in combination with the other one covers everything.  This Coccinelle script can fix any stragglers:
```
@@
type T;
identifier A;
@@
 T A[
-0
 ];
```

    spatch --sp-file za.cocci --very-quiet --include-headers --dir .

pabigot commented:
@stephanosio can this be moved out of draft now, or are you waiting for #25481?

stephanosio commented:
> @stephanosio can this be moved out of draft now, or are you waiting for #25481?

As I anticipated, the problem isn't going to be as simple as simply replacing `[0]` to `[]`; there are many errors in the CI.

I will look into fixing them.

joerchan committed:
Bluetooth: host: Don't use struct with zero size

Don't use the ATT structs that has contains only a flexible array
member. This is not supported by C99 standard, only through GNU C
extension with zero length array.

Signed-off-by: Joakim Andersson <joakim.andersson@nordicsemi.no>

joerchan commented:
@stephanosio I have fixed the ATT build issues, I pushed them and a rebased your branch to here: https://github.com/zephyrproject-rtos/zephyr/compare/master...joerchan:fix_zero_length_array
If you want I can force-push the branch to this PR.

I reverted the changes to ll_sw/pdu.h. They will require some larger changes.
Also there is one other instance in keys.h that is also not trivial to remove.

stephanosio commented:
> @stephanosio I have fixed the ATT build issues, I pushed them and a rebased your branch to here: [master...joerchan:fix_zero_length_array](https://github.com/zephyrproject-rtos/zephyr/compare/master...joerchan:fix_zero_length_array)
> If you want I can force-push the branch to this PR.
> 
> I reverted the changes to ll_sw/pdu.h. They will require some larger changes.
> Also there is one other instance in keys.h that is also not trivial to remove.

Thanks, feel free to do so.
